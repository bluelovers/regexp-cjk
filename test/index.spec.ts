import zhRegExp, { ParserEventEmitterEvent } from '../index';
import ParserEventEmitter from 'regexp-parser-event';

describe(`^(?:戰記)$`, () =>
{
	const source = `^(?:戰記)$`;
	const expected = /^\^\(\?:\[[^\[\]]{2,}\]\[[^\[\]]{2,}\]\)\$$/;

	test(`base`, () =>
	{

		let actual = new zhRegExp(source, '', {
			greedyTable: true,
		});

		expect(actual.source).toMatch(expected);
		expect(actual.source).not.toStrictEqual(source);

		//expect(actual).toBeInstanceOf(Date);
		expect(actual).toMatchSnapshot();

	});

	test(`event.on`, () =>
	{

		let bool: boolean;

		let actual = new zhRegExp(source, '', {
			greedyTable: true,
			on: {
				[ParserEventEmitterEvent.change](ast, event)
				{
					/*
					console.dir({
						ast,
						event,
					});
					 */

					expect({
						event,
						ast,
					}).toMatchSnapshot();

					bool = true;
				},
			},
		});

		expect(actual.source).toMatch(expected);
		expect(actual.source).not.toStrictEqual(source);

		expect(bool).toBeTruthy();

		//expect(actual).toBeInstanceOf(Date);
		expect(actual).toMatchSnapshot();

	});

})

describe(`check Punctuation`, () =>
{

	const actual = new zhRegExp(/([\p{Punctuation}])/u);
	const expected = /([\p{Punctuation}])/u;

	test(`base`, function ()
	{
		expect(actual.source).toStrictEqual(expected.source)
		;
		expect(actual).toMatchSnapshot();
	});

	test(`，`, function ()
	{
		let str = `，`;

		expect(actual.test(`，`)).toStrictEqual(expected.test(str));
		expect(actual.test(`，`)).toMatchSnapshot();
	});

})

describe(`simple test regex is not break`, () =>
{

	[
		`麻#_@_#痹`,
		`国國`,
	].forEach(v => _simpleTest002(v));

	[
		'(^|[^\\w\'’])(\\d+(?:,\\d+)*)(?![\\w\'’])',
		/国|國/.source,
	].forEach(v => _simpleTest001(v));

	function _simpleTest001(expected: string, cb?: (data: {
		expected: string,
		actual,
		r1,
		r2,
		ev,
	}) => any)
	{
		it(expected, async function ()
		{
			let ev = ParserEventEmitter.create(expected, 'u');
			ev.resume();

			let actual = ev.getSource(true);

			expect(actual).toStrictEqual(expected);

			let r1 = new zhRegExp(expected, 'u');
			let r2 = new zhRegExp(actual, 'u');

			/*
			console.dir({
				actual,
				r1,
				r2,
			});
			 */

			expect(r1.source)
				.toStrictEqual(r2.source)
			;

			expect(actual).toMatchSnapshot();
			expect(ev).toMatchSnapshot();

			if (cb)
			{
				await cb({
					expected,
					actual,
					r1,
					r2,
					ev,
				})
			}

		});

	}

	function _simpleTest002(expected: string)
	{
		_simpleTest001(expected, ({
			expected,
			actual,
			r1,
			r2,
			ev,
		}) => {
			expect(expected).toMatch(r1);
			expect(expected).toMatch(r2);
			expect(r1.source).not.toEqual(expected);
			expect(r2.source).not.toEqual(expected);

			expect(actual).toMatchSnapshot();
		});
	}

});

test(`not change`, function ()
{
	//console.log('it:inner', currentTest.title);
	//console.log('it:inner', currentTest.fullTitle());

	let bool: boolean;

	let actual = new zhRegExp(`111`, '', {
		greedyTable: true,
		on: {
			[ParserEventEmitterEvent.change](ast, event)
			{
				bool = true;
			}
		}
	});

	expect(bool).toBeFalsy();
	expect(actual).toMatchSnapshot();
});
