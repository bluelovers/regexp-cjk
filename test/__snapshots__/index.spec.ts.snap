// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`^(?:戰記)$ base 1`] = `/\\^\\(\\?:\\[战戦戰\\]\\[記记\\]\\)\\$/`;

exports[`^(?:戰記)$ event.on 1`] = `
Object {
  "ast": Object {
    "changed": true,
    "end": 6,
    "old_raw": "戰",
    "parent": Object {
      "changed": true,
      "elements": Array [
        [Circular],
        Object {
          "end": 7,
          "parent": [Circular],
          "raw": "記",
          "start": 6,
          "type": "Character",
          "value": 35352,
        },
      ],
      "end": 8,
      "parent": Object {
        "changed": true,
        "elements": Array [
          Object {
            "end": 2,
            "kind": "start",
            "parent": [Circular],
            "raw": "^",
            "start": 1,
            "type": "Assertion",
          },
          [Circular],
          Object {
            "end": 9,
            "kind": "end",
            "parent": [Circular],
            "raw": "$",
            "start": 8,
            "type": "Assertion",
          },
        ],
        "end": 9,
        "parent": Object {
          "changed": true,
          "end": 10,
          "flags": Object {
            "dotAll": false,
            "end": 10,
            "global": false,
            "ignoreCase": false,
            "multiline": false,
            "parent": [Circular],
            "raw": "",
            "start": 10,
            "sticky": false,
            "type": "Flags",
            "unicode": false,
          },
          "parent": null,
          "pattern": [Circular],
          "raw": "/^(?:戰記)$/",
          "start": 0,
          "type": "RegExpLiteral",
        },
        "raw": "^(?:戰記)$",
        "start": 1,
        "type": "Pattern",
      },
      "raw": "(?:戰記)",
      "start": 2,
      "type": "Group",
    },
    "raw": "[战戦戰]",
    "start": 5,
    "type": "Character",
    "value": 25136,
  },
  "event": "change",
}
`;

exports[`^(?:戰記)$ event.on 2`] = `
Object {
  "ast": Object {
    "changed": true,
    "end": 7,
    "old_raw": "記",
    "parent": Object {
      "changed": true,
      "elements": Array [
        Object {
          "changed": true,
          "end": 6,
          "old_raw": "戰",
          "parent": [Circular],
          "raw": "[战戦戰]",
          "start": 5,
          "type": "Character",
          "value": 25136,
        },
        [Circular],
      ],
      "end": 8,
      "parent": Object {
        "changed": true,
        "elements": Array [
          Object {
            "end": 2,
            "kind": "start",
            "parent": [Circular],
            "raw": "^",
            "start": 1,
            "type": "Assertion",
          },
          [Circular],
          Object {
            "end": 9,
            "kind": "end",
            "parent": [Circular],
            "raw": "$",
            "start": 8,
            "type": "Assertion",
          },
        ],
        "end": 9,
        "parent": Object {
          "changed": true,
          "end": 10,
          "flags": Object {
            "dotAll": false,
            "end": 10,
            "global": false,
            "ignoreCase": false,
            "multiline": false,
            "parent": [Circular],
            "raw": "",
            "start": 10,
            "sticky": false,
            "type": "Flags",
            "unicode": false,
          },
          "parent": null,
          "pattern": [Circular],
          "raw": "/^(?:戰記)$/",
          "start": 0,
          "type": "RegExpLiteral",
        },
        "raw": "^(?:戰記)$",
        "start": 1,
        "type": "Pattern",
      },
      "raw": "(?:戰記)",
      "start": 2,
      "type": "Group",
    },
    "raw": "[記记]",
    "start": 6,
    "type": "Character",
    "value": 35352,
  },
  "event": "change",
}
`;

exports[`^(?:戰記)$ event.on 3`] = `/\\^\\(\\?:\\[战戦戰\\]\\[記记\\]\\)\\$/`;

exports[`check Punctuation base 1`] = `/\\(\\[\\\\p\\{Punctuation\\}\\]\\)/u`;

exports[`check Punctuation ， 1`] = `true`;

exports[`not change 1`] = `/111/`;

exports[`simple test regex is not break (^|[^\\w'’])(\\d+(?:,\\d+)*)(?![\\w'’]) 1`] = `"(^|[^\\\\w'’])(\\\\d+(?:,\\\\d+)*)(?![\\\\w'’])"`;

exports[`simple test regex is not break (^|[^\\w'’])(\\d+(?:,\\d+)*)(?![\\w'’]) 2`] = `
ParserEventEmitter {
  "_events": Object {
    "change": [Function],
  },
  "_eventsCount": 1,
  "_maxListeners": undefined,
  "astRegExpLiteral": Object {
    "end": 38,
    "flags": Object {
      "dotAll": false,
      "end": 38,
      "global": false,
      "ignoreCase": false,
      "multiline": false,
      "parent": [Circular],
      "raw": "u",
      "start": 37,
      "sticky": false,
      "type": "Flags",
      "unicode": true,
    },
    "parent": null,
    "pattern": Object {
      "changed": false,
      "elements": Array [
        Object {
          "changed": false,
          "elements": Array [
            Object {
              "alternatives": Array [
                Array [
                  Object {
                    "changed": true,
                    "end": 3,
                    "kind": "start",
                    "parent": [Circular],
                    "raw": "^",
                    "start": 2,
                    "type": "Assertion",
                  },
                ],
                Array [
                  Object {
                    "changed": false,
                    "elements": Array [
                      Object {
                        "changed": true,
                        "end": 8,
                        "kind": "word",
                        "negate": false,
                        "parent": [Circular],
                        "raw": "\\\\w",
                        "start": 6,
                        "type": "CharacterSet",
                      },
                      Object {
                        "changed": true,
                        "end": 9,
                        "parent": [Circular],
                        "raw": "'",
                        "start": 8,
                        "type": "Character",
                        "value": 39,
                      },
                      Object {
                        "changed": true,
                        "end": 10,
                        "parent": [Circular],
                        "raw": "’",
                        "start": 9,
                        "type": "Character",
                        "value": 8217,
                      },
                    ],
                    "end": 11,
                    "negate": true,
                    "old_raw": "[^\\\\w'’]",
                    "parent": [Circular],
                    "raw": "[^\\\\w'’]",
                    "start": 4,
                    "type": "CharacterClass",
                  },
                ],
              ],
              "changed": false,
              "end": 11,
              "old_raw": "^|[^\\\\w'’]",
              "parent": [Circular],
              "raw": "^|[^\\\\w'’]",
              "start": 2,
              "type": "Disjunction",
            },
          ],
          "end": 12,
          "name": null,
          "old_raw": "(^|[^\\\\w'’])",
          "parent": [Circular],
          "raw": "(^|[^\\\\w'’])",
          "references": Array [],
          "start": 1,
          "type": "CapturingGroup",
        },
        Object {
          "changed": false,
          "elements": Array [
            Object {
              "changed": false,
              "element": Object {
                "changed": true,
                "end": 15,
                "kind": "digit",
                "negate": false,
                "parent": [Circular],
                "raw": "\\\\d",
                "start": 13,
                "type": "CharacterSet",
              },
              "end": 16,
              "greedy": true,
              "max": Infinity,
              "min": 1,
              "old_raw": "+",
              "parent": [Circular],
              "raw": "\\\\d+",
              "start": 15,
              "type": "Quantifier",
            },
            Object {
              "changed": false,
              "element": Object {
                "changed": false,
                "elements": Array [
                  Object {
                    "changed": true,
                    "end": 20,
                    "parent": [Circular],
                    "raw": ",",
                    "start": 19,
                    "type": "Character",
                    "value": 44,
                  },
                  Object {
                    "changed": false,
                    "element": Object {
                      "changed": true,
                      "end": 22,
                      "kind": "digit",
                      "negate": false,
                      "parent": [Circular],
                      "raw": "\\\\d",
                      "start": 20,
                      "type": "CharacterSet",
                    },
                    "end": 23,
                    "greedy": true,
                    "max": Infinity,
                    "min": 1,
                    "old_raw": "+",
                    "parent": [Circular],
                    "raw": "\\\\d+",
                    "start": 22,
                    "type": "Quantifier",
                  },
                ],
                "end": 24,
                "old_raw": "(?:,\\\\d+)",
                "parent": [Circular],
                "raw": "(?:,\\\\d+)",
                "start": 16,
                "type": "Group",
              },
              "end": 25,
              "greedy": true,
              "max": Infinity,
              "min": 0,
              "old_raw": "*",
              "parent": [Circular],
              "raw": "(?:,\\\\d+)*",
              "start": 24,
              "type": "Quantifier",
            },
          ],
          "end": 26,
          "name": null,
          "old_raw": "(\\\\d+(?:,\\\\d+)*)",
          "parent": [Circular],
          "raw": "(\\\\d+(?:,\\\\d+)*)",
          "references": Array [],
          "start": 12,
          "type": "CapturingGroup",
        },
        Object {
          "changed": false,
          "elements": Array [
            Object {
              "changed": false,
              "elements": Array [
                Object {
                  "changed": true,
                  "end": 32,
                  "kind": "word",
                  "negate": false,
                  "parent": [Circular],
                  "raw": "\\\\w",
                  "start": 30,
                  "type": "CharacterSet",
                },
                Object {
                  "changed": true,
                  "end": 33,
                  "parent": [Circular],
                  "raw": "'",
                  "start": 32,
                  "type": "Character",
                  "value": 39,
                },
                Object {
                  "changed": true,
                  "end": 34,
                  "parent": [Circular],
                  "raw": "’",
                  "start": 33,
                  "type": "Character",
                  "value": 8217,
                },
              ],
              "end": 35,
              "negate": false,
              "old_raw": "[\\\\w'’]",
              "parent": [Circular],
              "raw": "[\\\\w'’]",
              "start": 29,
              "type": "CharacterClass",
            },
          ],
          "end": 36,
          "kind": "lookahead",
          "negate": true,
          "old_raw": "(?![\\\\w'’])",
          "parent": [Circular],
          "raw": "(?![\\\\w'’])",
          "start": 26,
          "type": "Assertion",
        },
      ],
      "end": 36,
      "old_raw": "(^|[^\\\\w'’])(\\\\d+(?:,\\\\d+)*)(?![\\\\w'’])",
      "parent": [Circular],
      "raw": "(^|[^\\\\w'’])(\\\\d+(?:,\\\\d+)*)(?![\\\\w'’])",
      "start": 1,
      "type": "Pattern",
    },
    "raw": "/(^|[^\\\\w'’])(\\\\d+(?:,\\\\d+)*)(?![\\\\w'’])/u",
    "start": 0,
    "type": "RegExpLiteral",
  },
  Symbol(kCapture): false,
}
`;

exports[`simple test regex is not break 国|國 1`] = `"国|國"`;

exports[`simple test regex is not break 国|國 2`] = `
ParserEventEmitter {
  "_events": Object {
    "change": [Function],
  },
  "_eventsCount": 1,
  "_maxListeners": undefined,
  "astRegExpLiteral": Object {
    "end": 6,
    "flags": Object {
      "dotAll": false,
      "end": 6,
      "global": false,
      "ignoreCase": false,
      "multiline": false,
      "parent": [Circular],
      "raw": "u",
      "start": 5,
      "sticky": false,
      "type": "Flags",
      "unicode": true,
    },
    "parent": null,
    "pattern": Object {
      "changed": false,
      "elements": Array [
        Object {
          "alternatives": Array [
            Array [
              Object {
                "changed": true,
                "end": 2,
                "parent": [Circular],
                "raw": "国",
                "start": 1,
                "type": "Character",
                "value": 22269,
              },
            ],
            Array [
              Object {
                "changed": true,
                "end": 4,
                "parent": [Circular],
                "raw": "國",
                "start": 3,
                "type": "Character",
                "value": 22283,
              },
            ],
          ],
          "changed": false,
          "end": 4,
          "old_raw": "国|國",
          "parent": [Circular],
          "raw": "国|國",
          "start": 1,
          "type": "Disjunction",
        },
      ],
      "end": 4,
      "old_raw": "国|國",
      "parent": [Circular],
      "raw": "国|國",
      "start": 1,
      "type": "Pattern",
    },
    "raw": "/国|國/u",
    "start": 0,
    "type": "RegExpLiteral",
  },
  Symbol(kCapture): false,
}
`;

exports[`simple test regex is not break 国國 1`] = `"国國"`;

exports[`simple test regex is not break 国國 2`] = `
ParserEventEmitter {
  "_events": Object {
    "change": [Function],
  },
  "_eventsCount": 1,
  "_maxListeners": undefined,
  "astRegExpLiteral": Object {
    "end": 5,
    "flags": Object {
      "dotAll": false,
      "end": 5,
      "global": false,
      "ignoreCase": false,
      "multiline": false,
      "parent": [Circular],
      "raw": "u",
      "start": 4,
      "sticky": false,
      "type": "Flags",
      "unicode": true,
    },
    "parent": null,
    "pattern": Object {
      "changed": false,
      "elements": Array [
        Object {
          "changed": true,
          "end": 2,
          "parent": [Circular],
          "raw": "国",
          "start": 1,
          "type": "Character",
          "value": 22269,
        },
        Object {
          "changed": true,
          "end": 3,
          "parent": [Circular],
          "raw": "國",
          "start": 2,
          "type": "Character",
          "value": 22283,
        },
      ],
      "end": 3,
      "old_raw": "国國",
      "parent": [Circular],
      "raw": "国國",
      "start": 1,
      "type": "Pattern",
    },
    "raw": "/国國/u",
    "start": 0,
    "type": "RegExpLiteral",
  },
  Symbol(kCapture): false,
}
`;

exports[`simple test regex is not break 国國 3`] = `"国國"`;

exports[`simple test regex is not break 麻#_@_#痹 1`] = `"麻#_@_#痹"`;

exports[`simple test regex is not break 麻#_@_#痹 2`] = `
ParserEventEmitter {
  "_events": Object {
    "change": [Function],
  },
  "_eventsCount": 1,
  "_maxListeners": undefined,
  "astRegExpLiteral": Object {
    "end": 10,
    "flags": Object {
      "dotAll": false,
      "end": 10,
      "global": false,
      "ignoreCase": false,
      "multiline": false,
      "parent": [Circular],
      "raw": "u",
      "start": 9,
      "sticky": false,
      "type": "Flags",
      "unicode": true,
    },
    "parent": null,
    "pattern": Object {
      "changed": false,
      "elements": Array [
        Object {
          "changed": true,
          "end": 2,
          "parent": [Circular],
          "raw": "麻",
          "start": 1,
          "type": "Character",
          "value": 40635,
        },
        Object {
          "changed": true,
          "end": 3,
          "parent": [Circular],
          "raw": "#",
          "start": 2,
          "type": "Character",
          "value": 35,
        },
        Object {
          "changed": true,
          "end": 4,
          "parent": [Circular],
          "raw": "_",
          "start": 3,
          "type": "Character",
          "value": 95,
        },
        Object {
          "changed": true,
          "end": 5,
          "parent": [Circular],
          "raw": "@",
          "start": 4,
          "type": "Character",
          "value": 64,
        },
        Object {
          "changed": true,
          "end": 6,
          "parent": [Circular],
          "raw": "_",
          "start": 5,
          "type": "Character",
          "value": 95,
        },
        Object {
          "changed": true,
          "end": 7,
          "parent": [Circular],
          "raw": "#",
          "start": 6,
          "type": "Character",
          "value": 35,
        },
        Object {
          "changed": true,
          "end": 8,
          "parent": [Circular],
          "raw": "痹",
          "start": 7,
          "type": "Character",
          "value": 30201,
        },
      ],
      "end": 8,
      "old_raw": "麻#_@_#痹",
      "parent": [Circular],
      "raw": "麻#_@_#痹",
      "start": 1,
      "type": "Pattern",
    },
    "raw": "/麻#_@_#痹/u",
    "start": 0,
    "type": "RegExpLiteral",
  },
  Symbol(kCapture): false,
}
`;

exports[`simple test regex is not break 麻#_@_#痹 3`] = `"麻#_@_#痹"`;
